mcelog plugin

krzysztofx opened the pull request:
Plugin for collecting machine check exceptions  reported by [mcelog ](http://www.mcelog.org/)
The plugin at current stage does the following:
• Checks mcelog server liveliness, reports a failure if it’s not running of if it fails.
• Retrieve Memory Corrected and Uncorrected Errors from mcelog Unix socket (client protocol)
• Publishes collected data as notifications and statistics(values)

rubenk commented:
Thanks for your contribution @krzysztofx!
I'll leave some comments inline.

rubenk commented on README:
```diff
@@ -926,6 +931,11 @@ Prerequisites
      `varnish' plugin.
      <http://varnish-cache.org>
 
+  * mcelog
```
I don't think you need to add this block in this section, since it is mcelog is not a library and needed for building collectd.

krzysztofx replied:
right, it's not compilation prerequisite

rubenk commented on README:
```diff
@@ -198,14 +198,19 @@ Features
       Motherboard sensors: temperature, fan speed and voltage information,
       using mbmon(1).
 
+    - mcelogmon
```
The plugin is called mcelog

krzysztofx replied:
I'll fix that

rubenk commented on configure.ac:
```diff
@@ -6314,6 +6314,7 @@ AC_PLUGIN([match_regex],         [yes],                     [The regex match])
 AC_PLUGIN([match_timediff],      [yes],                     [The timediff match])
 AC_PLUGIN([match_value],         [yes],                     [The value match])
 AC_PLUGIN([mbmon],               [yes],                     [Query mbmond])
+AC_PLUGIN([mcelog],              [yes],                     [Machine Check Exceptions notifications])
```
This always builds and installs mcelog, but I think it is Linux-specific?

krzysztofx replied:
right, I'll fix that

rubenk commented on src/Makefile.am:
```diff
@@ -669,6 +669,16 @@ mbmon_la_LIBADD += -lsocket
+mcelog_la_SOURCES = mcelog.c
+mcelog_la_LDFLAGS = $(PLUGIN_LDFLAGS)
+mcelog_la_LIBADD =
+if BUILD_WITH_LIBSOCKET
```
I don't think you need this, it is mainly for Solaris iirc, and mcelog seems Linux-only.

krzysztofx replied:
libsocket linkage removed

octo commented on src/mcelog.c:
```diff
@@ -0,0 +1,431 @@
+        break;
+      }
+      if ((poll_fd.revents & POLLIN) || (poll_fd.revents & POLLPRI)) {
+        if ((p_file = fdopen(dup(g_mcelog_config.sock_fd), "r")) != NULL) {
```
More very long blocks, deeply nested. Please handle the error, not the positive case and remove levels of indentation. The parsing should probably go into its own function to make it more readable.

krzysztofx replied:
ok, I will change it

octo commented on src/mcelog.c:
```diff
@@ -0,0 +1,431 @@
+          while (fgets(buf, sizeof(buf), p_file)) {
+            /* Got empty line or "done" */
+            if ((strlen(buf) == 1) ||
+                ((strlen(buf) >= 5) && (!strcmp(buf, "done\n")))) {
```
The `strlen(buf) >= 5` comparison is unnecessary.

octo commented on src/mcelog.c:
```diff
@@ -0,0 +1,431 @@
+          memset(buf, 0, sizeof(buf));
+          while (fgets(buf, sizeof(buf), p_file)) {
+            /* Got empty line or "done" */
+            if ((strlen(buf) == 1) ||
```
I think `strcmp("\n", buf) == 0` is easier to read (and more strict) than `strlen(buf) == 1`

krzysztofx replied:
I will fix it

octo commented on src/mcelog.c:
```diff
@@ -0,0 +1,431 @@
+                memset(&memory_record, 0, sizeof(memory_record));
+              }
+            }
+            if (strlen(buf) >= 5) {
```
```c
if (strlen(buf) < 5)
  continue;
```

krzysztofx replied:
Ok understood. First handle negative scenario flow

octo commented on src/mcelog.c:
```diff
@@ -0,0 +1,431 @@
+                DEBUG("%s: done", MCELOG_PLUGIN);
+                break;
+              }
+              if (!memcmp(buf, socket_str, strlen(socket_str))) {
```
This has a high risk of reading uninitialized memory from `buf`. Please use `strncmp()` instead.

krzysztofx replied:
ok

octo commented on src/mcelog.c:
```diff
@@ -0,0 +1,431 @@
+      pthread_exit(NULL);
+    }
+    pthread_mutex_unlock(&mcelog_thread_lock);
+    int poll_ret = poll(&poll_fd, 1, POLL_TIMEOUT);
```
It might be worth sending this thread a signal (using `pthread_kill()` from the `shutdown()` callback) instead of busy looping here.

krzysztofx replied:
You mean adding signal handler to thread function  poll_worker ? Or just terminating thread asynchronously sending SIGTERM ?  I have noticed that some plugins use that approach but  I think this is not applicable here since we have to do some cleanup like calling fclose and so on.

krzysztofx replied:
I will terminate thread with thread with pthread_cancel and put cleanup code into callback function  set by pthread_cancel_push. Is it ok from your POV ?

octo commented on src/mcelog.c:
```diff
@@ -0,0 +1,431 @@
+  mcelog_thread_shutdown = 1;
+  pthread_mutex_unlock(&mcelog_thread_lock);
+  pthread_join(g_mcelog_config.tid, NULL);
+  pthread_mutex_destroy(&mcelog_thread_lock);
```
Please don't destroy mutexes that you initialized with the static initializer.

krzysztofx replied:
ok

octo commented on src/mcelog.c:
```diff
@@ -0,0 +1,431 @@
+  pthread_mutex_unlock(&mcelog_thread_lock);
+  pthread_join(g_mcelog_config.tid, NULL);
+  pthread_mutex_destroy(&mcelog_thread_lock);
+  shutdown(g_mcelog_config.sock_fd, 0);
```
Please check that `sock_fd` is in fact a valid file descriptor before calling `shutdown()`.

krzysztofx replied:
ok

octo commented on src/mcelog.c:
```diff
@@ -0,0 +1,431 @@
+static _Bool mcelog_thread_shutdown = 0;
+static pthread_mutex_t mcelog_thread_lock = PTHREAD_MUTEX_INITIALIZER;
+
+static const char socket_str[] = "SOCKET";
```
Having these `const char[]` up here doesn't really improve the parsing code below. Please simply use string literals in the worker thread.

krzysztofx replied:
I will change it

krzysztofx commented:
@octo @rubenk   thanks for your  reviews.  I'll respond  your comments soon.

krzysztofx commented on src/mcelog.c:
```diff
@@ -0,0 +1,431 @@
+typedef struct mcelog_config_s mcelog_config_t;
+typedef struct mcelog_memory_rec_s mcelog_memory_rec_t;
+
+mcelog_config_t g_mcelog_config;
```
I'd like to provide way to reinitalize   g_mcelog_config struct, so  how do you see usage of compound literals there ? I mean:
```c
static mcelog_config_t g_mcelog_config ;$
// skippded lines
static void mcelog_config_init_default(void) {$
  g_mcelog_config = (mcelog_config_t){$
    .logfile = "/var/log/mcelog",$
    .unix_sock = {$
      .sun_family = AF_UNIX,$
      .sun_path = "/var/run/mcelog-client",$
    },$
    .tid = 0,$
    .sock_fd = -1,$
  };$
```

krzysztofx commented on src/mcelog.c:
```diff
@@ -0,0 +1,431 @@
+    pthread_mutex_unlock(&mcelog_thread_lock);
+    int poll_ret = poll(&poll_fd, 1, POLL_TIMEOUT);
+
+    if (poll_ret > 0) {
```
ok

krzysztofx commented on src/mcelog.c:
```diff
@@ -0,0 +1,431 @@
+  vl.values = &(value_t){.derive = (derive_t)mr.corrected_err_total};
+  plugin_dispatch_values(&vl);
+
+  snprintf(type_instance, DATA_MAX_NAME_LEN, "corrected_memory_errors_in_%s",
```
I'll fix it

krzysztofx commented on src/mcelog.c:
```diff
@@ -0,0 +1,431 @@
+              (struct sockaddr *)&(g_mcelog_config.unix_sock),
+              sizeof(struct sockaddr_un)) < 0) {
+    sstrerror(errno, errbuff, sizeof(errbuff));
+    ERROR("%s: Failed to connect to mcelog server. %s", MCELOG_PLUGIN, errbuff);
```
ok

krzysztofx commented:
Hi @octo @rubenk 
I've pushed patch addressing your comments and also fix for build failure on  epel5 setup.
Please let me know if  you will find out  anything else  that could be improved.

rubenk commented on README:
```diff
@@ -202,10 +202,10 @@ Features
-      and reported to software) reported by mcelog and generate appropriate
-      notifications when machine check errors are detected.
+    - mcelog
+      Logs and accounts machine check exceptions (in particular memory, IO, and
```
I liked the original text better. The plugin doesn't actually log MCE's right?

krzysztofx replied:
It logs only the  amount of corrected/uncorrected  memory related errors  per  memory SOCKET and CHANNEL.  No MCE error codes are logged if that's what you mean. Data source is equivalent to output of  'mcelog --client', for example:
```sh
[root@silpixa00392165 kmatczax]# mcelog --client
Memory errors
SOCKET 0 CHANNEL any DIMM any
corrected memory errors:
        0 total
        0 in 24h
uncorrected memory errors:
        8 total
        8 in 24h

SOCKET 0 CHANNEL 0 DIMM any
corrected memory errors:
        0 total
        0 in 24h
uncorrected memory errors:
        8 total
        8 in 24h
```

krzysztofx commented:
Hi

I've committed  fix for typedef redefinition. All check passed but now I see merge conflict with no details given.  Please let me know if I could do anything to fix that, or if any other changes in plugin  code are still required.

krzysztofx commented:
@rubenk @octo 
Would you mind If I rebase that PR against master branch to fix the conflict ?

krzysztofx commented:
Rebase done

krzysztofx commented:
Hi 
@rubenk @octo 
Sorry for pinging , but is there anything that could be done to get that plugin merged? If yes, please let me know. 
I'm developing extension for that plugin that collects  all MCE exceptions, so it would be easier to base on already merged version.

rubenk commented:
Could you re-order or squash some of your commits? You now have multiple commits with the same commit messages.

rubenk commented on src/Makefile.am:
```diff
@@ -677,6 +677,13 @@ mbmon_la_LIBADD += -lsocket
+pkglib_LTLIBRARIES += mcelog.la
+mcelog_la_SOURCES = mcelog.c
+mcelog_la_LDFLAGS = $(PLUGIN_LDFLAGS)
+mcelog_la_LIBADD =
```
this is not needed.

rubenk commented on src/collectd.conf.pod:
```diff
@@ -3322,6 +3322,28 @@ TCP-Port to connect to. Defaults to B<411>.
 
+=head2 Plugin C<mcelog>
+
+The C<mcelog plugin> uses mcelog to retrieve machine check exceptions, etc.
```
What does the 'etc.' stand for? It doesn't add much information, I'd either leave it out or extend the sentence.

rubenk commented on src/collectd.conf.pod:
```diff
@@ -3322,6 +3322,28 @@ TCP-Port to connect to. Defaults to B<411>.
+mcelog server is running. When the server is running, the plugin will tail the
+specified logfile to retrieve machine check exception information and send a
+notification with the details from the logfile. The plugin will use the mcelog
+client protocol to retrieve memory errors.
```
I'd prefer 'retrieve machine check exceptions' instead of 'memory errors'.

krzysztofx replied:
At the moment mcelog's client/server protocol  returns only memory related MCE's, see http://www.mcelog.org/protocol.html. That's why I used 'memory errors' term. Would you like to extend it to  all MCE,  or maybe  use term 'memory related machine check exceptions' ?

rubenk commented on src/collectd.conf.pod:
```diff
@@ -3322,6 +3322,28 @@ TCP-Port to connect to. Defaults to B<411>.
+=over 4
+
+=item B<McelogClientSocket> I<Path>
+Connect to mcelog client socket using the UNIX domain socket at I<Path>.
```
Connect to the mcelog client socket

rubenk commented on src/collectd.conf.pod:
```diff
@@ -3322,6 +3322,28 @@ TCP-Port to connect to. Defaults to B<411>.
+
+=item B<McelogLogfile> I<Path>
+
+mcelog file to parse. Defaults to B<"/var/log/mcelog">.
```
The mcelog logfile to parse.

krzysztofx commented:
@rubenk 
Thanks for your review. I'll squash redundant commits  and address your comments soon

Thanks
Krzysztof

rubenk commented on src/mcelog.c:
```diff
@@ -0,0 +1,518 @@
+
+  /* synchronization via write lock since sock_fd may be changed here */
+  pthread_rwlock_wrlock(&self->lock);
+  self->sock_fd = socket(PF_UNIX, SOCK_STREAM, 0);
```
Could you use SOCK_CLOEXEC here?

rubenk replied:
Oh, and while you're at it, add SOCK_NONBLOCK to the flags, so you don't have to do the fcntl later.

krzysztofx replied:
regarding SOCK_NONBLOCK please see [d212c7f](https://github.com/collectd/collectd/pull/2003/commits/d212c7f2fb8af6d3ade42aaf1d019c5a2d7e8d3b). It failed on epel5

rubenk replied:
Ah right, sorry for the confusion. The situation has changed now though, after we released 5.7 we dropped support for RHEL5 from master. Jenkins is still doing build on RHEL5 but those will be cleaned up shortly and you can ignore them.

rubenk commented on src/mcelog.c:
```diff
@@ -0,0 +1,518 @@
+
+  /* Set socket timeout option */
+  if (setsockopt(self->sock_fd, SOL_SOCKET, SO_SNDTIMEO,
+                 (char *)&socket_timeout, sizeof(socket_timeout)) < 0)
```
I think the (char *) cast is unneeded.

rubenk commented on src/mcelog.c:
```diff
@@ -0,0 +1,518 @@
+static void *poll_worker(__attribute__((unused)) void *arg) {
+  char errbuf[MCELOG_BUFF_SIZE];
+  mcelog_thread_running = 1;
+  FILE **pp_file = calloc(1, sizeof(FILE *));
```
Please use sizeof on the var, not the type, so sizeof(*pp_file)

rubenk commented on src/mcelog.c:
```diff
@@ -0,0 +1,518 @@
+      continue;
+
+    mcelog_memory_rec_t memory_record;
+    memset(&memory_record, 0, sizeof(memory_record));
```
Please use designated initialiser here, so `mcelog_memory_rec_t memory_record = {0}`

krzysztofx commented:
My apologies for squashing by mistake all previous  commits into [c8f08229](https://github.com/collectd/collectd/pull/2003/commits/c8f08229b48085338394a99847b7c16de252b0da). I hope it doesn't make it much harder to review. If you think I could somehow revert that rebase and squash only specific commits please let me know. 
All recent comments have been addressed in [6016e1d](https://github.com/collectd/collectd/pull/2003/commits/6016e1d7ac73b9ac0e0eb58350f9ee710ccda4a5l)

Now I see merge conflict with the master branch. Do you agree for rebase ? 

Thanks
Krzysztof

krzysztofx commented:
[pull-requests-github_trigger-aggregation](https://ci.collectd.org/job/pull-requests-build-on-jessie-amd64-with-default-toolchain/1379/) failed due to:
 ```
mcelog.c:437:5: error: missing braces around initializer [-Werror=missing-braces]
     mcelog_memory_rec_t memory_record = {0};
```

Looks like this job has more restrictive compiler settings than others . Should I put additional braces like {{0}} or ignore that fail ?

rubenk commented:
No, that's not it, the compiler settings are the same. This job is the first that runs and the other ones are only run if this one succeeds.

This looks like https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53119.

Perhaps you could try changing the order of the members of the mcelog_memory_rec_s struct around to work around it? My guess is that this won't happen if the first member is an int instead of char[]

krzysztofx commented:
@rubenk 
Changing order of struct members helped for missing braces warning. Now as expected  I got failures on epel5 builds due to unknown socket flags:
 
```
mcelog.c: In function 'socket_reinit':
mcelog.c:163: error: 'SOCK_CLOEXEC' undeclared (first use in this function)
mcelog.c:163: error: (Each undeclared identifier is reported only once
mcelog.c:163: error: for each function it appears in.)
mcelog.c:163: error: 'SOCK_NONBLOCK' undeclared (first use in this function)
make[3]: *** [mcelog.lo] Error 1
```

So according to previous comments this is acceptable, right ? 
Please let me know if I should rebase to master. 

thanks
Krzysztof

rubenk commented:
Thanks @krzysztofx, yes, these errors on EPEL5 are perfectly alright.
Please rebase and then I'll merge.

krzysztofx commented:
@rubenk 
Rebase done

Thanks
Krzysztof
